docs/tutorials/cpp.md

```diff
@@ -1,43 +1,75 @@
-        - `provider_fs` for loading data from the file system (`ICU4XDataProvider::create_fs()`). This also requires enabling a syntax on the `icu_provider` crate.
-    - Set `CARGO_PROFILE_RELEASE_LTO=true` to enable link-time optimization
-    - Set `CARGO_PROFILE_RELEASE_OPT_LEVEL="s"` to optimize for size
-    - See [cargo profiles](cargo-profiles) for more options
- - Copy the header files from the `icu_capi` crate
-    - If you're using the default crate registry, they will be in `~/.cargo/registry/src/*/icu_capi-1.0.0/cpp/include`
-    - If you cannot find the crate, download it from crates.io
-    - Make sure to use the same version as above
-        - While we selected a version for `icu_capi_staticlib` before, and here we're looking at `icu_capi`, those versions will always be in sync. You can sanity check this by running `cargo pkgid -p icu_capi`.
- - Compile with `g++ -std=c++17 icu/target/release/libicu_capi_staticlib.a -ldl -lpthread -lm`. C++ versions beyond C++17 are supported, as are other C++ compilers.
+## Building ICU4X
+
+After [installing Rust](https://rustup.rs), create a local Rust crate to configure ICU4X:
+
+```shell
+cargo init icu --lib
+cd icu
+echo "[lib]" >> Cargo.toml
+echo "crate-type = ['cdylib', 'staticlib']" >> Cargo.toml
+```
+
+Now, add the `icu_capi` crate with the desired features:
+
+```shell
+cargo add icu_capi --features foo
+echo "extern crate icu_capi;" > src/lib.rs
+```
+
+Features include:
+- `compiled_data` \[default\] to include data (`ICU4XDataProvider::create_compiled()`)
+- `simple_logger` \[default\] enable basic stdout logging of error metadata. Further loggers can be added on request.
+- `default_components` \[default\] activate all stable ICU4X components. For smaller builds, this can be disabled, and components can be added with features like `icu_list`.
+- `buffer_provider` for working with blob data providers (`ICU4XDataProvider::create_from_byte_slice()`)
+
+Default features can be turned off with
+```
+cargo add icu_capi --no-default-features
+```
+
+You can now build the library:
+
+```shell
+cargo build --release
+cd ..
+```
+
+- Be sure to pass `--release` to get an optimized build
+- Set `CARGO_PROFILE_RELEASE_LTO=true` to enable link-time optimization
+- Set `CARGO_PROFILE_RELEASE_OPT_LEVEL="s"` to optimize for size
+- See [cargo profiles](cargo-profiles) for more options
+
+You should now have `icu/target/release/libicu.a` and `icu/target/release/libicu.{so,dll,dylib}`, ready to compile into your C++ binary.
+
+## Obtaining the header files
+
+The header files are shipped inside the crate, which Cargo has put somewhere on your system. You can find its location with
+```shell
+cargo metadata --manifest-path icu/Cargo.toml --format-version 1 | jq '.packages[] | select(.name == "icu_capi").manifest_path' | xargs dirname
+```
+
+Copy them to your working directory:
+```shell
+cp -r $(!!)/cpp/include .
```

Comment:
Nit: this is a lesson in ICU4X, not a lesson in shell scripting. I had to look up what `$(!!)` does. I much prefer using plain variables where possible. For example, save the result of the previous command into a variable like `ICU_CAPI_DIR` and then reference it here.

Reply:
It works and they can copy paste it. I don't want to introduce variables because those don't work in make files for example. `$(!!)` also doesn't work, but in the history it gets replaced with the full invocation, which can then be used. The one-liner is just too long to be understandable here.